### PR TITLE
Add system.cpu.physical.count and system.cpu.logical.count metrics

### LINKED ISSUE #98: add system.cpu.count metric
Details: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22099

### PR DESCRIPTION
**Pull request by frzifus**
ref: https://github.com/open-telemetry/opentelemetry-specification/issues/3556
Closes #98

### CODE REVIEW

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (frzifus) [inline specification/metrics/semantic_conventions/system-metrics.md:42]: @dmitryax, @mx-psi you may want to have a look.

Comment (mx-psi) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: While I understand the practical benefit of having this as a metric (as mentioned on https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22099#issuecomment-1581682604), I still feel like if we abide strictly by the data model this is more of a resource attribute: a fixed (at least, as fixed as `os.version` or `host.image.version` can be) property of a resource that generates telemetry. I don't have a strong opinion on whether practical concerns should trump the data model purity argument, but I think we should still consider both.

For Datadog's use case a resource attribute would be better, but I don't think our use case is very common (we intend to map this information to a vendor-specific payload at the exporter level and a resource attribute allows for statelessness in deployments while if we have this information as metrics we need all metrics to end up on the same payload).

Comment (mx-psi) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: The discussion on open-telemetry/opentelemetry-specification/pull/2384 is also relevant. This comment https://github.com/open-telemetry/opentelemetry-specification/pull/2384#issuecomment-1056063450 makes me think that my thumb rule of "if it has a fixed value over time then it's not a metric" makes sense.

@trask @bogdandrutu @tigrannajaryan, any thoughts on this PR and the metric vs. resource attribute discussion?

Comment (frzifus) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: > if it has a fixed value over time then it's not a metric

But is it a fixed value? My I am not an expert on that field, but my understanding is that it can change.
At least this [random VMware article](https://www.bdrsuite.com/blog/what-is-vmware-hot-add-hot-plug-how-will-it-impact-numa/) claims it too. ^^

Comment (chambridge) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: I'll also share this scenario as public clouds allow you to resize an instance the cpu count can certainly change over the lifetime of a system: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-resize.html

Comment (mx-psi) [inline specification/metrics/semantic_conventions/system-metrics.md:42]: I think we should clarify if this is the number of physical or logical processors

Comment (mx-psi) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: I guess we probably want to report both, or at least account for the possibility of wanting to do so in the future, so we may want to make the name reflective of which one of the two we are reporting (so maybe having `physical` on the name somewhere makes sense, unless it's clear that when people talk about CPUs they are talking about physical ones?))

Comment (frzifus) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: hm.. what would you recommend? Something like `system.cpu.vcount` and `system.cpu.ncount`?

cc @chambridge

Comment (mx-psi) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: Sounds like I am the only one arguing for resource attribute, I suppose we can mark this discussion as solved

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (mx-psi) [inline specification/metrics/semantic_conventions/system-metrics.md:43]: `machine-id` sounds like a resource attribute to me (it is the value of `host.id` for non-containerized, non-cloud systems after all https://github.com/open-telemetry/semantic-conventions/blob/main/specification/resource/semantic_conventions/host.md) . I think we should leave the `Attribute values` list empty for these two.

Since we do not specify the resource attributes added to these metrics here, we can discuss whether we should add `host.id` to the metrics generated by the hostmetrics receiver on the Collector repository.

cc @dmitryax any thoughts on this?

**Review by mx-psi — COMMENTED** [review]
Just two suggestions, otherwise looks good to me

**Review by mx-psi — APPROVED** [review]

**Comment by frzifus** [comment]
cc @specs-semconv-approvers @open-telemetry/specs-semconv-maintainers

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (ChrsMark) [inline specification/metrics/semantic_conventions/system-metrics.md:42]: Would that make sense to add the `physical`/`virtual` suffixes as attribute values instead?
Also what are the possible attribute values for the `device` Attribute Key?

Comment (jamesmoessis) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: What's the use case for splitting up physical/virtual? Additionally, is it possible for instrumentation to reliably detect if the CPU is physical or virtual? My initial thoughts are I think I would prefer it if this was just one metric.

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (jamesmoessis) [reply specification/metrics/semantic_conventions/system-metrics.md]: > I think we should clarify if this is the number of physical or logical processors

@mx-psi can you clarify what the use case of this is? As a user of the `hostmetricsreceiver` anecdotally wouldn't need this to be split up. From an instrumentation standpoint, is it possible to detect if CPUs are virtual or physical?

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (dmitryax) [reply specification/metrics/semantic_conventions/system-metrics.md]: I'm not sure what `device` is and why it's needed as an attribute for this metric while other CPU metrics don't have it. Sound redundant to me. And `host.id` is a resource attribute defined in the conventions and added by resourcedetection processor.

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (mx-psi) [reply specification/metrics/semantic_conventions/system-metrics.md:42]: > What's the use case for splitting up physical/virtual? 

At a minimum, we should clarify which one of the two we are adding as a semantic convention, since the term can be ambiguous. I think both are helpful to understand the hardware a machine has, in our (Datadog's) case, it would be helpful to have both to provide feature parity with our existing vendor-specific instrumentation.

>  is it possible for instrumentation to reliably detect if the CPU is physical or virtual?

Widely used libraries such as psutil allow [fetching this information](https://psutil.readthedocs.io/en/latest/#psutil.cpu_count) on all major operating systems (Linux, Windows, macOS, *BSD, AIX...). For example, on Linux it is possible to fetch this information so long as the process has access to `/proc/cpuinfo` and `/sys/devices/`. This is a similar level of access to that required for fetching the information related to other system metrics already in the spec.

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (mx-psi) [reply specification/metrics/semantic_conventions/system-metrics.md]: Alright, I think we are in agreement, since the `device` attribute was removed we can mark this one as resolved cc @frzifus

Comment (frzifus) [reply specification/metrics/semantic_conventions/system-metrics.md]: oh yes, but it seems I made a mistake in the last commit https://github.com/open-telemetry/semantic-conventions/pull/99/commits/d01a81ae2f62d1da3a643b577fe99945dbfe8feb. Let me fix that.

Comment (frzifus) [reply specification/metrics/semantic_conventions/system-metrics.md]: fixed it.

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (tigrannajaryan) [reply specification/metrics/semantic_conventions/system-metrics.md]: I think we need a definition of (or a link to) what "virtual CPU count" is. The link to psutil talks about "logical" cpu count, not virtual.

Are we trying to capture the nuances of virtualized environments? For a system running on a VM what does each of these metrics return? Is "physical" count the actual physical count or the number allocated by the hypervisor to the VM?

Comment (mx-psi) [reply specification/metrics/semantic_conventions/system-metrics.md]: We started talking about "physical" vs "logical" on the original discussion (https://github.com/open-telemetry/semantic-conventions/pull/99#discussion_r1228223112 ) but ended talking about physical vs virtual and I failed to realize that this was confusing and could be interpreted as virtual in the sense of a VM.

To me, **the important distinction is between physical** (number of CPUs available in the machine) **and logical** (physical multiplied by the number of threads that can run on each CPU through hyperthreading). This is the way `psutil` defines it:
>  “logical CPUs” means the number of physical cores multiplied by the number of threads that can run on each core (this is known as Hyper Threading)

see also [this SE question](https://unix.stackexchange.com/a/88290) or the Wikipedia article on [Hyperthreading](https://en.wikipedia.org/wiki/Hyper-threading).

The terms "virtual" and "logical" are used interchangeably sometimes in this context (e.g. see the first section on the Wikipedia article linked above) but virtual can also mean in a VM in common parlance, so based on the above discussion I feel like logical is a less ambiguous term.

On a virtualized environment the 'physical' number may not correspond to the actual number of CPUs the machine has depending on what the hypervisor allocated, but I still think physical is the common name used for this. 

Since the physical and logical definitions do not depend on the hypervisor I think it may be clearer now that these numbers are available to instrumentation provided they can read the appropriate system files/call the right APIs.

@tigrannajaryan @jamesmoessis does this make the definitions clearer to you?

Comment (tigrannajaryan) [reply specification/metrics/semantic_conventions/system-metrics.md]: I think we should avoid using "virtual" and stick to "logical". The word "virtual" appears only twice on wikipedia article and "logical" is mentioned 15 times. That will help avoid confusion for virtualized environments.

If we stick to "physical" and "logical" I don't think we need to strictly define the behavior for VMs. The numbers should be "whatever the guest OS sees".

Comment (dmitryax) [reply specification/metrics/semantic_conventions/system-metrics.md]: I believe the most common use case for these metrics on the backends would be to calculate cpu utilization from `system.cpu.time` metric. We should make it clear which one of them should be used for that purpose.

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (dmitryax) [inline specification/metrics/semantic_conventions/system-metrics.md:43]: Should the unit be `{cpu}`? AFAIR `1` is recommended for ratio metrics

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (tigrannajaryan) [reply specification/metrics/semantic_conventions/system-metrics.md]: To add to what @dmitryax said. I think we need to be clear which of these new metrics' count is supposed to be the total number of values you see for `cpu` attribute of `system.cpu.time` metric. There must be an exact correspondence conceptually between what `system.cpu.time`/`cpu` dimension describes and what one of these 2 new metrics describes.

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (frzifus) [reply specification/metrics/semantic_conventions/system-metrics.md]: changed it, thanks!

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (mx-psi) [reply specification/metrics/semantic_conventions/system-metrics.md]: `system.cpu.time`, at least on psutil/gopsutil and (therefore in the current implementation of the hostmetrics receiver) reports the time for each **logical** CPU:

<details>
<summary>
Go and Python short programs showing that the logical CPUs are used for <code>system.cpu.time</code>
</summary>

Short check using `psutil`:

```python
#! /usr/bin/python
import psutil
times = psutil.cpu_times(percpu=True)
logical_count = psutil.cpu_count(logical=True)

print(f"logical count: {logical_count} system.cpu.time number of CPUs: {len(times)}")
```

**Output:**
```
❯ ./main.py
logical count: 20 system.cpu.time number of CPUs: 20
```

Go program using `gopsutil` (the library used by the host metrics receiver):

```go
package main

import (
	"fmt"
	"github.com/shirou/gopsutil/v3/cpu"
)

func main(){
       // fn used by hostmetrics receiver here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3dfc0f9/receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go#L39
	times, err := cpu.Times(true)
	if err != nil {
		panic(err)
	}

	logicalCount, err := cpu.Counts(true)
	if err != nil {
		panic(err)
	}
	
	fmt.Printf("logical count: %d system.cpu.time number of CPUs: %d\n", logicalCount, len(times))
}
```

**Output:**
```
❯ go run main.go
logical count: 20 system.cpu.time number of CPUs: 20
```
</details>

We can clarify this on the description of `system.cpu.time`.

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (mx-psi) [inline specification/metrics/semantic_conventions/system-metrics.md:40]: From https://github.com/open-telemetry/semantic-conventions/pull/99#discussion_r1246483912, we can clarify here that this is the number of logical CPUs.

```suggestion
| system.cpu.utilization    | Difference in system.cpu.time since the last measurement, divided by the elapsed time and number of logical CPUs | 1     | Gauge                                             | Double     | state            | idle, user, system, interrupt, etc. |
```

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (mx-psi) [inline specification/metrics/semantic_conventions/system-metrics.md:38]: From https://github.com/open-telemetry/semantic-conventions/pull/99#discussion_r1246483912, here is a proposed description to clarify that this metrics measures time spent per **logical** CPU.

```suggestion
| system.cpu.time           | Seconds each logical CPU spent on each mode                                                              | s     | Counter                                           | Double     | state            | idle, user, system, interrupt, etc. |
```

**Comment by mx-psi** [comment]
@tigrannajaryan @jamesmoessis the metrics have been renamed and the description of `system.cpu.time` and `system.cpu.utilization` now clarify that there is one dimension per logical CPU. PTAL!

---

Separate note, this will likely conflict with #154

**Comment by frzifus** [comment]
@mx-psi conflics are resolved.

@tigrannajaryan @jamesmoessis whats your thoughts?

**Comment by frzifus** [comment]
oh shoot, it seems the last force push revered some recommendations form @mx-psi . Let me check.

**Comment by frzifus** [comment]
@tigrannajaryan can you check - added the comments again.

File: specification/system/system-metrics.md
Comment (tigrannajaryan) [inline specification/system/system-metrics.md:39]:
```suggestion
|                           |                                                                                                                  |       |                                                   |            | cpu              | Logical CPU number [0..n-1]                 |
```

File: specification/system/system-metrics.md
Comment (tigrannajaryan) [inline specification/system/system-metrics.md:41]:
```suggestion
|                           |                                                                                                                  |       |                                                   |            | cpu              | Logical CPU number (0..n)                   |
```

**Review by tigrannajaryan — COMMENTED** [review]
I think this can work. Just left suggestions on how to make it even more clear.

**Comment by frzifus** [comment]
rebased again

**Review by jamesmoessis — APPROVED** [review]
Thanks @mx-psi for clarifying physical/virtual/logical and @frzifus for the PR. Looks all good to me now 👍

**Review by tigrannajaryan — APPROVED** [review]

**Comment by frzifus** [comment]
resolved changelog conflict

**Comment by frzifus** [comment]
@open-telemetry/specs-semconv-maintainers is something missing?

**Comment by mx-psi** [comment]
Looks like a link broke after the reorganization
```
  ERROR: 1 dead links found!
  [✖] file:///home/runner/work/semantic-conventions/semantic-conventions/docs/general/metrics-general.md#instrument-types → Status: 400
```

**Comment by dmitryax** [comment]
It's still unclear from the description what's the difference between logical/physical and what should be used to calculate time/utilization per core (@mx-psi said that it's logical). I think that important information to cover in the description

**Comment by frzifus** [comment]
What do you have in mind @dmitryax - Something like this on top?
```
Logical CPU refers to the virtual cores created by the operating system to manage multitasking, while a physical CPU represents the actual physical processor core(s) on the hardware.
```

> ... what should be used to calculate time/utilization per core 

Isnt that covered here?
```
system.cpu.time ... Seconds each logical CPU spent on each mode
system.cpu.utilization ... Difference in system.cpu.time since the last measurement, divided by the elapsed time and number of logical CPUs
```

**Comment by dmitryax** [comment]
> Something like this on top?

Yes, this sounds descriptive enough.

> Isnt that covered here?

I missed that. SGTM

**Comment by frzifus** [comment]
added that part @dmitryax  :)

File: docs/system/system-metrics.md
Comment (dmitryax) [inline docs/system/system-metrics.md:45]: I was thinking about putting it into the description instead of a note. let me know WDYT
```suggestion
| system.cpu.physical.count | Reports the number of available physical CPUs (actual physical processor cores on the hardware)                                             | {cpu} | UpDownCounters                                    | Int64      |                  |                                     |
| system.cpu.logical.count  | Reports the number of available logical CPUs (virtual processor cores created by the operating system to manage multitasking)                                              | {cpu} | UpDownCounters                                    | Int64      |                  |                                     |
```

**Comment by frzifus** [comment]
Works for me too. I rephrased it a bit. Can you check one more time @dmitryax ? ^^

File: docs/system/system-metrics.md
Comment (dmitryax) [inline docs/system/system-metrics.md:44]: I would use `logical` term more as it's referred to in other metrics' descriptions.
```suggestion
| system.cpu.logical.count  | Reports the number of logical (virtual) processor cores created by the operating system to manage multitasking | {cpu} | UpDownCounters                                    | Int64      |                  |                                     |
```

**Review by dmitryax — APPROVED** [review]

**Comment by mx-psi** [comment]
@open-telemetry/specs-semconv-maintainers anything left to do on this PR?

**Comment by mx-psi** [comment]
@frzifus 

```
/home/runner/work/semantic-conventions/semantic-conventions/docs/system/system-metrics.md:36: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

**Comment by frzifus** [comment]
Fixed

**Comment by jsuereth** [comment]
Ideally, we'd like this change to either be generated off YAML *or* wait until the system metrics are defined in YAML, then merge this PR with that change.

File: docs/system/system-metrics.md
Comment (arminru) [inline docs/system/system-metrics.md:43]:
```suggestion
|                           |                                                                                                                  |       |                                                   |            | cpu              | Logical CPU number (0..n)           |
| system.cpu.physical.count | Reports the number of actual physical processor cores on the hardware | {cpu} | UpDownCounter                                     | Int64      |                  |                                     |
| system.cpu.logical.count  | Reports the number of logical (virtual) processor cores created by the operating system to manage multitasking | {cpu} | UpDownCounter                                     | Int64      |                  |                                     |
```

(this would not be an issue after https://github.com/open-telemetry/semantic-conventions/pull/99#issuecomment-1660510502)

**Review by arminru — APPROVED** [review]

File: specification/metrics/semantic_conventions/system-metrics.md
Comment (rockdaboot) [reply specification/metrics/semantic_conventions/system-metrics.md]: > I'll also share this scenario as public clouds allow you to resize an instance the cpu count can certainly change over the lifetime of a system: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-resize.html

Sorry to comment on a closed PR, but that user guide clearly says that you need to change your instance type to change the number of CPUs. This means turning off the old instance and start up a new one, including IP address change.

Similar to @mx-psi, I'd argue for CPU count being a resource attribute. Is it possible to have CPU count as metric and resource attribute?

**Pull request merged**
